### ISSUE TITLE
Module templates variables issues

### ISSUE THREAD

**Issue by webgear** [contao/core#7933]
Hi there,

We are having some issues with modules and templates. Even with a clean install of Contao (no custom themes, no extra modules other than the core ones etc), modules are not being rendered (except for the text/html modules and other static ones). For the newsletter module, the template is not rendered at all. We've also created a custom module and tried to use the compile() method of the module to pass data to a template.

The variables are being set, but cannot be used. If we try to ECHO the variables, nothing gets displayed, not even the HTML container. However, if we place an exit; instruction right before the $this->endblock() call, then the thing gets rendered properly. Same goes for the newsletter module. If we place the exit call, then the form shows up. If we remove it, nothing is shown.

Why is this happening and how can we fix it? This is so weird and with the lack of documentation (developer documentation), the thing doesn't get easier to learn.

The configuration that we've done is:
- Layout -> Themes -> Frontend modules -> Module type = Subscribe (Newsletter)
- Articles -> New article -> New element of type Module , and the module is the one we created in the Theme section

Thanks in advance.

**Comment by discordier** [comment]
- You have added a layout in your theme I suppose and attached that one to the page root?
- What exactly do you get in the frontend? Is there anything mentioned in the log?
- What software are you using in which version? PHP? apache? nginx?

We need plenty of more details to be able to be of help.

**Comment by webgear** [comment]
- yes, the layout is not an issue; without a layout, an error is thrown and no page is rendered, so that is not the problem
- frontend HTML output for the newsletter subscription module (only the headline is rendered):

```<div class="mod_article last block" id="subscribe">
      <h1>Subscribe</h1>
  <p class=""></p>
</div>```

- software: 2 boxes: windows xampp installation with Apache 2.4, PHP 5.6.8, as well as a CentOS 6.5 x64 system with Apache 2.2, PHP 5.4.3 (production machine)

The logs don't output anything. I've specifically deleted the files and ran the pages where the modules don't show up, and no error is logged. Additionally, when intentionally triggering an error (eg: calling an undefined method in the template), the error is correctly logged.

I must again iterate the fact that putting an exit instruction before the $this->endblock() line in system/modules/newsletter/templates/newsletter/nl_default.html5 causes the module frontend to render correctly, and subsequently, var_dump instructions followed by exit are correctly printing the values (which are not nulls). Also, calls to $this->showTemplateVars() and $this->dumpTemplateVars() correctly print out the values set from the modules. However, when using the actual values (aka echo-ing them) without placing the exit or die instruction afterwards, the HTML is completely stripped from the output. This happens with ANY template variable, in ANY module, so it's not something specific to just one specific block.

**Comment by webgear** [comment]
After some more checks, in trying to locate the actual block start/end processing functions, I ended up in the system/modules/core/library/Contao/BaseTemplate.php file. There are four ob_end_clean() calls in that file, replacing them with ob_end_flush() will cause the page output to be displayed as soon as the buffer is processed. The elements on the page are showing up 5 times !, but only once they have the correct contents (meaning the subscription form is rendered). However, that only time occurs in the middle of all the block calls, so naturally they are discarded (because of the ob_end_clean() calls). I don't know how to approach this, could this be caused by bad page/article design or something else?

In the backend, I have created:
- Site structure -> Website root node which has a child "Home" which is a Regular page
- under Content -> Articles -> I have the homepage which has 2 child nodes: first is a Text widget, the other is a Module of type Newsletter Subscribe. I've attached a couple of screenshots to better illustrate that.

![site-structure](https://cloud.githubusercontent.com/assets/94492/8813945/11c2c1b6-3010-11e5-93c3-91775c9d50c7.jpg)
![frontend-modules](https://cloud.githubusercontent.com/assets/94492/8813944/11c03eaa-3010-11e5-84aa-4a2aff08cd70.jpg)
![articles](https://cloud.githubusercontent.com/assets/94492/8813946/11c6450c-3010-11e5-9732-b732b3e15f93.jpg)
![newsletter-module](https://cloud.githubusercontent.com/assets/94492/8813947/11c9dd20-3010-11e5-9bd5-1a7e0ce312ae.jpg)
![newsletter-module-config](https://cloud.githubusercontent.com/assets/94492/8813948/11c9db90-3010-11e5-8fe0-8bea216cd38e.jpg)

**Comment by xchs** [comment]
Did you already run the [Contao check](https://docs.contao.org/books/manual/3.5/en/01-installation/configuring-the-live-server.html#the-contao-check)?

**Comment by webgear** [comment]
This is totally awkward. Am i the only one that can't do anything but place static blocks in the system?

![contao-file-perms](https://cloud.githubusercontent.com/assets/94492/8820794/2dd4d398-3061-11e5-8798-b8ea31d1f868.jpg)
![contao-check](https://cloud.githubusercontent.com/assets/94492/8820795/2df0ed80-3061-11e5-95aa-23247a6578f1.jpg)

**Comment by BugBuster1701** [comment]
Have you started the "Validate an existing installation"? This checks if the files not corrupted.

**Comment by webgear** [comment]
Yes. Nothing is broken except for the gitignore file.

Contao Check 10.6
Validate an installation

Version

Found a Contao 3.5.0 installation.

Corrupt files

.gitignore

Your installation is not up to date.

**Comment by webgear** [comment]
There was an unknown issue with the page, I guess. I created a new, empty page, in which I only included the newsletter subscribe module. Oddly enough that worked. Why it did not work in the first try on the clean installation is beyond me. Now the other module (custom) also works correctly, variables are being passed correctly and shown. I believe this ticket can be closed now, but maybe it it something worth investigating, because its not normal behavior.
